README.txt
Cell Matrix - Implementation of Conway's Game of Life Algorithm

Project site:
http://gatc.ca/projects/cell-matrix/


Dependencies:
Python 2.7+ (https://www.python.org/)
Pygame 1.9.1 (https://www.pygame.org/)

Usage:
Interphase module (http://gatc.ca/projects/interphase/)
>GUI interface module.
PyJ2D 0.27_dev (http://gatc.ca/projects/pyj2d/) / Jython 2.2.1+ (http://www.jython.org/)
>Optional to port Pygame app to Java environment of JVM 6.0+ (https://www.java.com).
Pyjsdl 0.21_dev (http://gatc.ca/projects/pyjsdl/) / Pyjs 0.8.1_dev (http://www.pyjs.org/)
>Optional to port Pygame app to JS environment of Web browser.


Instructions:
Cell Matrix runs with Python and the Pygame library. Input of Game of Life patterns can be done from the interface panel by the controls LOAD from a customized pattern.txt file and GET from the system clipboard. The app is run with the command 'python cell.py'. Alternatively, the app can run in the Java environment using Jython and the PyJ2D library with the command 'jython cell.py' or 'java -jar jython.jar cell.py'. The app can also run in the JavaScript environment using Pyjs compilation and the Pyjsdl library with the command '[pyjs_path]/bin/pyjsbuild -O cell.py --dynamic-link -o output' and copying data folder to pyjsbuild output.

Controls
Scroll up (UP/KP8)
Scroll down (DOWN/KP2)
Scroll left (LEFT/KP4)
Scroll right (RIGHT/KP6)
Reset (r)
Clear (c)
Pause/Edit toggle (Escape/e)
Edit cell (LMouse)
Panel toggle (p)
Pattern-panel/clipboard (LMouse/RMouse)


Released under the GPL3 license (http://www.gnu.org/licenses/gpl.html).



CHANGELOG.txt
1.0     2017-04-20
-update matrix reset to turn off edit mode.
-add icon for the system display.
-update for py3 compatibility.
-update interphase module.
-update app for alternate framework.

0.9     2015-09-05

-initial release



cell.py
#!/usr/bin/env python

"""
Cell Matrix - Implementation of Conway's Game of Life Algorithm
Copyright (C) 2015 James Garnon

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

Cell Matrix version 1.0
Download Site: http://gatc.ca
"""

from __future__ import division
import os, sys, math, random
if os.name in ('posix', 'nt', 'os2', 'ce', 'riscos'):
    import pygame
    platform = 'pc'
elif os.name == 'java':
    import pyj2d
    sys.modules['pygame'] = pyj2d
    pygame = pyj2d
    platform = 'jvm'
    if sys.version_info[0:2] == (2,2):
        from sets import Set
else:
    import pyjsdl as pygame
    platform = 'js'

import interphase
interphase.init(pygame)

import env
env.engine = pygame
env.interface = interphase
env.platform = platform
env.size = 8
env.color0 = pygame.Color(0,0,0)
env.color1 = pygame.Color(0,255,0)

from control import Control
from interface import MatrixInterface


class Matrix:

    def __init__(self, width, height):
        self.width, self.height = width, height
        self.size = env.size
        self.color0 = env.color0
        self.color1 = env.color1
        self.rect = pygame.Rect(0, 0, self.size-2, self.size-2)
        self.cell0 = pygame.Surface((self.size-2, self.size-2))
        self.cell0.fill(self.color0)
        self.cell1 = pygame.Surface((self.size-2, self.size-2))
        self.cell1.fill(self.color1)
        self.setPool = []
        self.listPool = []
        self.patterns = {}
        self.pattern_id = 0
        self.screen, self.background, self.control, self.panel, self.clock = self.setup(width,height)
        if self.panel.matrix:
            psize = self.panel.get_size()
            ppos = self.panel.get_position()
            self.panel_clear1 = pygame.Surface(psize)
            self.panel_pos1 = pygame.Rect((ppos[0]-(psize[0]//2),ppos[1]-(psize[1]//2)), psize)
            self.panel_clear1.blit(self.screen, (0,0), self.panel_pos1)     #pyjsdl blit - IndexSizeError - clip area
            self.panel_clear2 = pygame.Surface((psize[0],12))
            self.panel_pos2 = pygame.Rect((ppos[0]-(psize[0]//2),ppos[1]-(psize[1]//2)+88), (psize[0],12))
            self.panel_clear2.blit(self.screen, (0,0), self.panel_pos2)
            self.panel_pos = ppos[1]
            self.panel_display = True
        else:
            self.panel_display = False
        self.display_update = False
        if env.platform in ('pc','jvm'):
            self.xrange = list(range((width//10), width-(width//10)))
            self.yrange = list(range((height//10), height-(height//10)))
        else:
            self.xrange = list(range((width//5), width-(width//5)))
            self.yrange = list(range((height//5), height-(height//5)))
        self.view = [0,0]
        self.scroll = False
        self.grid = {}
        self.grid_cache = {}
        self.cell_seed()
        self.nb = []
        for i in (-1, 0, 1):
            for j in (-1, 0, 1):
                if i == 0 and j == 0:
                    continue
                self.nb.append((i,j))
        self.quit = False
        self.pause = False
        self.edit = False
        self.change = False
        try:
            self.neighbor_set = set()
        except NameError:
            self.neighbor_set = Set()
            self.get_set = self.get_set_alt
        self.neighbors = {}
        self.neighbor_cache_init()
        self.ticks = 6
        self.tick = 0
        self.update_list = []

    def setup(self, w, h):
        screen = pygame.display.get_surface()
        background = pygame.Surface((w*self.size,h*self.size))
        background.fill((20,40,60))
        for x in range(w):
            for y in range(h):
                background.blit(self.cell0, ((x*self.size)+1, (y*self.size)+1))
        screen.blit(background, (0,0))
        pygame.display.flip()
        pygame.display.set_caption('Cell Matrix')
        icon = pygame.image.load(os.path.join('data', 'icon.png'))
        pygame.display.set_icon(icon)
        pygame.event.set_blocked(pygame.MOUSEMOTION)
        pygame.key.set_repeat(100,10)
        panel = MatrixInterface(self)
        panel._clipboard_init()
        control = Control(self, panel)
        clock = pygame.time.Clock()
        return screen, background, control, panel, clock

    def set_tick(self, value):
        ticks = {1:6, 2:3, 3:1, 4:0}[value]
        if self.ticks != ticks:
            self.ticks = ticks
            self.tick = self.ticks

    def get_set(self):
        if self.setPool:
            s = self.setPool.pop()
            s.clear()
            return s
        else:
            return set()

    def get_set_alt(self):
        if self.setPool:
            s = self.setPool.pop()
            s.clear()
            return s
        else:
            return Set()

    def get_list(self):
        if self.listPool:
            lst = self.listPool.pop()
            lst[:] = []
            return lst
        else:
            return []

    def edit_matrix(self):
        self.pause = True
        self.edit = True
        self.change = True

    def activate(self):
        self.edit = False
        self.pause = False

    def view_scroll(self, direction, view_change=1):
        self.view[0] += (direction[0]*view_change)
        self.view[1] += (direction[1]*view_change)
        self.change = True
        self.scroll = True

    def view_reset(self):
        self.view[0] = 0
        self.view[1] = 0
        self.change = True
        self.scroll = True

    def neighbor_cache_init(self):
        if env.platform in ('pc','jvm'):
            for x in range(-self.width//2, self.width+(self.width//2)):
                for y in range(-self.height//2, self.height+(self.height//2)):
                    self.neighbor_cache(x,y)
        else:
            for x in range(0, self.width):
                for y in range(0, self.height):
                    self.neighbor_cache(x,y)

    def cell_seed(self):
        for x in self.xrange:
            for y in self.yrange:
                if random.random() > 0.9:
                    self.grid[(x,y)] = None

    def reset_matrix(self):
        self.clear_matrix()
        self.cell_seed()
        self.tick = 0
        if self.edit:
            self.panel.edit_enable(False)
            self.activate()

    def clear_matrix(self):
        self.grid.clear()
        self.view[0] = 0
        self.view[1] = 0
        self.screen.blit(self.background, (0,0))
        pygame.display.flip()

    def clear_panel(self):
        self.panel_pos = self.panel.get_position()[1]
        if self.panel_pos != 441:
            self.screen.blit(self.panel_clear1, self.panel_pos1)
            self.update_list.append(self.panel_pos1)
        else:
            self.screen.blit(self.panel_clear2, self.panel_pos2)
            self.update_list.append(self.panel_pos2)

    def paste_pattern(self, x, y, pattern=None):
        if pattern is None:
            pattern = self.panel.get_clipboard()
            if not pattern:
                return
        pattern_start = False
        for line in pattern.split('\n'):
            if not line or self.check_pattern_comment(line):
                if not pattern_start:
                    continue
                else:
                    break
            else:
                pattern_start = True
            for i, char in enumerate(line):
                if char == '.':
                    pass
                elif char in ('O','*'):
                    self.grid[x+i,y] = None
            y += 1

    def check_pattern_comment(self, line):
        try:
            return line.startswith(('!','#','//'))
        except TypeError:
            for ch in ('!','#','//'):
                if line.startswith(ch):
                    return True
            else:
                return False

    def cell_add(self, x, y):
        self.grid[(x, y)] = None
        self.update_list.append(pygame.Rect(x*self.size, y*self.size, self.size, self.size))

    def cell_remove(self, x, y):
        if self.grid[(x,y)]:
            self.setPool.append(self.grid_cache[(x,y)])
        del self.grid[(x,y)]
        self.singlecell_clear((x,y))

    def cell_toggle(self, x, y):
        if (x,y) not in self.grid:
            self.cell_add(x,y)
        else:
            self.cell_remove(x,y)

    def neighbor_cache(self, x, y):
        self.neighbors[(x,y)] = self.neighbor_generate(x,y)

    def neighbor_generate(self, x, y):
        neighbors = self.get_list()
        for i,j in self.nb:
            neighbors.append((x+i,y+j))
        return neighbors

    def neighbor_locate(self, x, y):
        if (x,y) in self.neighbors:
            neighbor = self.get_set()
            for cell in self.neighbors[(x,y)]:
                if cell not in self.grid:
                    neighbor.add(cell)
            return neighbor
        else:
            neighbors = self.neighbor_generate(x,y)
            neighbor = self.get_set()
            for cell in neighbors:
                if cell not in self.grid:
                    neighbor.add(cell)
            return neighbor

    def neighbor_count(self, x, y):
        if (x,y) in self.neighbors:
            count = 0
            for cell in self.neighbors[(x,y)]:
                if cell in self.grid:
                    count += 1
            return count
        else:
            neighbors = self.neighbor_generate(x,y)
            count = 0
            for cell in neighbors:
                if cell in self.grid:
                    count += 1
            return count

    def grid_clear(self):
        for cell in self.grid_cache:
            self.setPool.append(self.grid_cache[cell])
        self.grid_cache.clear()

    def cell_live(self):
        for cell in self.grid:
            self.grid[cell] = self.neighbor_locate(cell[0],cell[1])
            if 1 < (8-len(self.grid[cell])) < 4:
                self.grid_cache[cell] = None
            self.neighbor_set.update(self.grid[cell])

    def cell_reprod(self):
        for cell in self.neighbor_set:
            if self.neighbor_count(cell[0],cell[1]) == 3:
                self.grid_cache[cell] = None
        self.neighbor_set.clear()

    def cell_growth(self):
        self.grid_clear()
        self.cell_live()
        self.cell_reprod()
        grid = self.grid
        self.grid = self.grid_cache
        self.grid_cache = grid

    def singlecell_clear(self, cell):
        x = self.view[0]
        y = self.view[1]
        if (x <= cell[0] < self.width+x) and (y <= cell[1] < self.height+y):
            self.rect.x = ((cell[0]-x)*self.size)+1
            self.rect.y = ((cell[1]-y)*self.size)+1
            self.update_list.append( self.screen.blit(self.cell0, self.rect) )

    def cell_clear_select(self, height):
        x = self.view[0]
        y = self.view[1]
        for cell in self.grid_cache:
            if (x <= cell[0] < self.width+x) and (y <= cell[1] < self.height+y):
                self.rect.x = ((cell[0]-x)*self.size)+1
                self.rect.y = ((cell[1]-y)*self.size)+1
                self.update_list.append( self.screen.blit(self.cell0, self.rect) )

    def cell_draw_select(self, height):
        x = self.view[0]
        y = self.view[1]
        for cell in self.grid:
            if (x <= cell[0] < self.width+x) and (y <= cell[1] < self.height+y):
                self.rect.x = ((cell[0]-x)*self.size)+1
                self.rect.y = ((cell[1]-y)*self.size)+1
                self.update_list.append( self.screen.blit(self.cell1, self.rect) )

    def cell_display_select(self):
        if not self.scroll:
            self.cell_clear_select(self.panel_pos)
            self.cell_draw_select(self.panel_pos)

    def cell_clear(self):
        x = self.view[0]
        y = self.view[1]
        for cell in self.grid_cache:
            if (x <= cell[0] < self.width+x) and (y <= cell[1] < self.height+y):
                self.rect.x = ((cell[0]-x)*self.size)+1
                self.rect.y = ((cell[1]-y)*self.size)+1
                self.update_list.append( self.screen.blit(self.cell0, self.rect) )

    def cell_draw(self):
        x = self.view[0]
        y = self.view[1]
        for cell in self.grid:
            if (x <= cell[0] < self.width+x) and (y <= cell[1] < self.height+y):
                self.rect.x = ((cell[0]-x)*self.size)+1
                self.rect.y = ((cell[1]-y)*self.size)+1
                self.update_list.append( self.screen.blit(self.cell1, self.rect) )

    def cell_display(self):
        if not self.scroll:
            self.cell_clear()
            self.cell_draw()
        else:
            self.screen.blit(self.background, (0,0))
            self.cell_draw()

    def update(self):
        self.update_list[:] = []
        if self.panel_display:
            self.clear_panel()
        self.panel.update()
        self.control.update()
        if not self.pause:
            if not self.tick:
                self.cell_growth()
                self.cell_display()
                self.tick = self.ticks
            else:
                self.tick -= 1
                if self.change:
                    self.cell_display()
                    self.change = False
                else:
                    if self.panel_display:
                        self.cell_display_select()
        else:
            if self.change:
                self.cell_display()
                self.change = False
            else:
                if self.panel_display:
                    self.cell_display_select()
        if self.panel_display:
            self.panel.draw(self.screen)
        if not self.scroll:
            pygame.display.update(self.update_list)
        else:
            pygame.display.update()
            self.scroll = False
        self.clock.tick(30)


def setup(w,h):
    pygame.init()
    pygame.display.set_mode((w*env.size,h*env.size))


def main():
    w, h = 60, 50
    setup(w,h)
    matrix = Matrix(w,h)
    while not matrix.quit:
        matrix.update()


matrix = None
def run():
    global matrix
    matrix.update()

def pre_run():
    global matrix
    matrix = Matrix(60, 50)
    pygame.display.setup(run)

def main_js():
    setup(60, 50)
    img = ['data/control_n.png', 'data/control_s.png', 'data/control_w.png',
            'data/control_e.png', 'data/button.png', 'data/icon.png']
    pygame.display.setup(pre_run, img)


if __name__ == '__main__':
    if platform != 'js':
        main()
    else:
        main_js()



control.py
#Cell Matrix - Copyright (C) 2015
#Released under the GPL3 License

from __future__ import division
import env
pygame = env.engine


class Control:

    def __init__(self, matrix, panel):
        self.matrix = matrix
        self.panel = panel
        self.dir = {pygame.K_UP:(0,-1), pygame.K_DOWN:(0,1), pygame.K_LEFT:(-1,0), pygame.K_RIGHT:(1,0)}
        self.dir_kp = {pygame.K_KP8:(0,-1), pygame.K_KP2:(0,1), pygame.K_KP4:(-1,0), pygame.K_KP6:(1,0)}

    def update(self):
        for event in pygame.event.get():
            if event.type == pygame.KEYDOWN:
                if event.key == pygame.K_r:
                    if self.matrix.edit:
                        self.matrix.activate()
                        self.panel.edit_enable(False)
                    self.matrix.reset_matrix()
                elif event.key == pygame.K_c:
                    self.matrix.clear_matrix()
                elif event.key in (pygame.K_e, pygame.K_ESCAPE):
                    if not self.matrix.edit:
                        self.matrix.edit_matrix()
                        self.panel.edit_enable(True)
                    else:
                        self.matrix.activate()
                        self.panel.edit_enable(False)
                elif event.key == pygame.K_p:
                    self.matrix.panel_display = not self.matrix.panel_display
                    self.matrix.clear_panel()
                elif event.key in (pygame.K_UP, pygame.K_DOWN, pygame.K_LEFT, pygame.K_RIGHT):
                    self.matrix.view_scroll(self.dir[event.key])
                elif event.key in (pygame.K_KP8, pygame.K_KP2, pygame.K_KP4, pygame.K_KP6):
                    self.matrix.view_scroll(self.dir_kp[event.key])
                elif event.key == pygame.K_q:
                    self.matrix.quit = True
            elif event.type == pygame.MOUSEBUTTONDOWN and not self.panel.rect.collidepoint(event.pos):
                if event.button == 1:
                    x = (event.pos[0]//env.size) + self.matrix.view[0]
                    y = (event.pos[1]//env.size) + self.matrix.view[1]
                    pattern = self.panel.get_pattern()
                    if pattern:
                        self.matrix.paste_pattern(x,y,pattern)
                    else:
                        self.matrix.cell_toggle(x,y)
                        self.matrix.change = True
                elif event.button == 3:
                    x = (event.pos[0]//env.size) + self.matrix.view[0]
                    y = (event.pos[1]//env.size) + self.matrix.view[1]
                    self.matrix.paste_pattern(x,y)
                    if self.matrix.edit:
                        self.matrix.change = True
            elif event.type == pygame.QUIT:
                self.matrix.quit = True
        return None



pattern.py
patterns = \
"""

//Game of Life Patterns
//http://www.conwaylife.com

//Cell Matrix program will load these patterns.
//This file can be edited without restrictions.
//Patterns added should have the format of either Life 1.05 or plaintext.
//Each pattern should be separated with at least a single blank line.
//Comment lines starting with // are ignored.
//Uses pattern.py if pattern.txt not available.


!Name: Glider
//!Author: Richard K. Guy
!The smallest, most common, and first discovered spaceship.
//!www.conwaylife.com/wiki/index.php?title=Glider
.O
..O
OOO


//#Life 1.05
#D Name: Garden of Eden 5
//#D Author: Nicolay Beluchenko
#D Currently the smallest-known Garden of Eden in Conway's Game of Life.
//#D www.conwaylife.com/wiki/index.php?title=Garden_of_Eden_5
//#N
//#P -5 -5
.***..**
.**.*.*.***
.***..*****
*.*.*.*.*.*
****.*.*.*
....***
.*.*.*.****
*.*.*.*.*.*
*****..***
***.*.*.**
...**..***



!Name: Gosper glider gun
//!Author: Bill Gosper
!The first known gun and the first known finite pattern with unbounded growth.
//!www.conwaylife.com/wiki/index.php?title=Gosper_glider_gun
........................O
......................O.O
............OO......OO............OO
...........O...O....OO............OO
OO........O.....O...OO
OO........O...O.OO....O.O
..........O.....O.......O
...........O...O
............OO


//#Life 1.05
#D Name: Dragon
//#D Author: Paul Tooke
#D The first c/6 spaceship to be constructed.
//#D www.conwaylife.com/wiki/index.php?title=Dragon
//#N
//#P -18 -9
............*
............**..............*
..........*.**.....*.*....**
.....*...*...***..*....*
**...*..*......*.*.....***..*
**...*.**......*...*.*.*
**...*..........*.*.......**
.....**..............*......*
.......*............*.*
.......*............*.*
.....**..............*......*
**...*..........*.*.......**
**...*.**......*...*.*.*
**...*..*......*.*.....***..*
.....*...*...***..*....*
..........*.**.....*.*....**
............**..............*
............*

"""


interface.py
#Cell Matrix - Copyright (C) 2015
#Released under the GPL3 License

from __future__ import division
import env
pygame = env.engine
interphase = env.interface


class MatrixInterface(interphase.Interface):

    def __init__(self, matrix):
        self.matrix = matrix
        interphase.Interface.__init__(self,
                position=((self.matrix.width*env.size)//2,(self.matrix.height*env.size)-47),
                color=(15,30,50), size=(350,96), control_minsize=(41,41),
                control_size='auto', control_response=100, moveable=False,
                position_offset=(0,88), font_color=(50,150,200),
                button_image='button.png', scroll_button='vertical')
        pygame.draw.rect(self.get_panel_image(True), (50,60,100), (162,28,178,48), 1)
        img = pygame.Surface((100,100), pygame.SRCALPHA)
        img.fill((50,50,50,150))
        self.set_control_image(surface=img)
        self.dir = {'N':(0,-1), 'S':(0,1), 'W':(-1,0), 'E':(1,0), 'NW':(-1,-1), 'NE':(1,-1), 'SW':(-1,1), 'SE':(1,1)}
        self.width, self.height = self.get_size()
        self.set_moveable()
        self.set_moveable('Fixed')
        self.panel_up = True
        if env.platform == 'js':
            self.add_controls_js()
            pygame.display.textbox_init()

    def add_controls(self):
        self.add(
            identity = 'Reset',
            control_type = 'control_toggle',
            position = (26,19),
            size = 'min_width',
            control_list = ['Reset'],
            tip_list = ['Reset matrix'],
            control_outline = True)
        self.add(
            identity = 'Clear',
            control_type = 'control_toggle',
            position = (26,39),
            size = 'min_width',
            control_list = ['Clear'],
            tip_list = ['Clear matrix'],
            control_outline = True)
        self.add(
            identity = 'Edit',
            control_type = 'control_toggle',
            position = (26,59),
            size = 'min_width',
            control_list = ['Edit'],
            tip_list = ['Edit matrix'],
            control_response = 0,
            control_outline = True,
            activated_toggle='lock')
        self.add(
            identity = 'PatternLoad',
            control_type = 'control_toggle',
            position = (235,86),
            size = 'auto_width',
            control_list = ['LOAD'],
            font_size = 8,
            tip_list = ['Retrieve from file'],
            control_response = 0,
            control_outline = True)
        self.add(
            identity = 'PatternGet',
            control_type = 'control_toggle',
            position = (265,86),
            size = 'auto_width',
            control_list = ['GET'],
            font_size = 8,
            tip_list = ['Retrieve from clipboard'],
            control_response = 0,
            control_outline = True)
        self.add(
            identity = 'Patterns',
            control_type = 'control_select',
            position = (250,18),
            size = (64,18),
            control_list = ['Patterns'],
            tip_list = ['Game of Life Patterns'],
            activated_toggle = False,
            control_response = 200,
            control_outline = True,
            split_text = False)
        self.add(
            identity = 'PatternInfo',
            control_type = 'textbox',
            position = (251,52),
            size = (176,46),
            color = (50,50,50),
            font_color = (255,255,255),
            font_type = 'courier',
            font_size = 8,
            control_list = [],
            tip_list = ['Pattern Information'],
            activated_toggle = False,
            control_outline = True)
        self.add(
            identity = 'Rate',
            control_type = 'control_select',
            position = (26,79),
            size = 'min_width',
            control_list = ['Rate 1','Rate 2','Rate 3','Rate 4'],
            tip_list = ['Processing rate'],
            split_text = False,
            control_response = 200,
            control_outline = True,
            reverse = True)
        for ident, tip, img, pos in ( ('N',['North'],['control_n.png'],(111,25)), ('S',['South'],['control_s.png'],(111,75)), ('W',['West'],['control_w.png'],(86,50)), ('E',['East'],['control_e.png'],(136,50)) ):
            self.add(
                identity = ident,
                control_type = 'control_toggle',
                position = pos,
                size = (25,25),
                control_list = [''],
                tip_list = tip,
                control_image = img,
                control_response = 25)
        for ident, tip, pos in ( ('NW',['North-West'],(86,25)), ('NE',['North-East'],(136,25)), ('SW',['South-West'],(86,75)), ('SE',['South-East'],(136,75)), ('X',['Center'],(111,50)) ):
            self.add(
                identity = ident,
                control_type = 'control_toggle',
                position = pos,
                size = (25,25),
                control_list = [''],
                tip_list = tip,
                control_response = 25)
        self.add(
            identity = '__TopFix',
            control_type = 'control_toggle',
            position = (self.get_size()[0]//2,4),
            size = (340,8),
            control_list = [''],
            control_image = 'none',
            control_response = 0,
            control_outline = False)
        self.add(
            identity = '__Fix',
            control_type = 'control_toggle',
            position = (324,86),
            color = (0,20,30),
            font_color = (0,120,160),
            font_size = 8,
            control_list = ['!'],
            control_response = 0,
            control_outline = True)
        self.add(
            identity = '__Help',
            control_type = 'control_toggle',
            position = (340,86),
            color = (0,20,30),
            font_color = (0,120,160),
            font_size = 8,
            control_list = ['?'],
            control_response = 0,
            control_outline = True)
        self.ctrl_patterns = self.get_control('Patterns')
        self.ctrl_patterninfo = self.get_control('PatternInfo')
        self.ctrl_patterninfo.set_active(False)
        self.ctrl_topfix = self.get_control('__TopFix')
        self.ctrl_topfix.set_enabled(False)

    def add_controls_js(self):
        self.add(
            identity = 'Textbox',
            control_type = 'control_toggle',
            position = (280,86),
            size = 'auto_width',
            control_list = ['v','^'],
            font_size = 8,
            tip_list = ['Toggle Textbox'],
            control_response = 0,
            control_outline = True)

    def load_pattern(self):
        if not env.platform == 'js':
            try:
                pattern_file = open('pattern.txt')
                patterns = pattern_file.read()
                pattern_file.close()
            except IOError:
                try:
                    from pattern import patterns
                except ImportError:
                    return
        else:
            try:
                from pattern import patterns
            except ImportError:
                return
        if patterns:
            self.register_pattern(patterns)

    def retrieve_pattern(self):
        pattern = self.get_clipboard()
        if pattern:
            self.register_pattern(pattern)

    def register_pattern(self, patterns):
        current_id = self.matrix.pattern_id
        pattern = []
        for line in patterns.split('\n'):
            if line.startswith('//'):
                continue
            if line.strip():
                pattern.append(line)
            else:
                if pattern:
                    self.set_pattern('\n'.join(pattern))
                    pattern[:] = []
        else:
            if pattern:
                self.set_pattern('\n'.join(pattern))
        if self.matrix.pattern_id == current_id:
            return
        if not self.ctrl_patterninfo.is_active():
            self.ctrl_patterninfo.set_active(True)
        if self.ctrl_patterns.is_activated():
            self.ctrl_patterns.set_activated(False)
        self.ctrl_patterns.set_tip(['Game of Life Patterns'])
        if current_id:
            self.ctrl_patterns.set_list_index(current_id)
        else:
            self.ctrl_patterns.reset()
        self.ctrl_patterninfo.set_value(self.matrix.patterns[int(self.ctrl_patterns.get_value().split(' ')[-1])])

    def set_pattern(self, pattern):
        self.matrix.pattern_id += 1
        self.matrix.patterns[self.matrix.pattern_id] = pattern
        if self.matrix.pattern_id > 1:
            self.ctrl_patterns.set_list(['Pattern '+str(self.matrix.pattern_id)], append=True)
        else:
            self.ctrl_patterns.set_list(['Pattern '+str(self.matrix.pattern_id)])

    def get_pattern(self):
        if self.matrix.patterns and self.ctrl_patterns.is_activated():
            return self.matrix.patterns[int(self.ctrl_patterns.value.split(' ')[-1])]
        else:
            return None

    def edit_enable(self, setting=True):
        ctrl = self.get_control('Edit')
        if setting:
            if not ctrl.is_activated():
                ctrl.set_activated(True)
        else:
            if ctrl.is_activated():
                ctrl.set_activated(False)

    def update(self):
        state = interphase.Interface.update(self)
        if state.control:
            if state.control in ('N', 'S', 'W', 'E', 'NW', 'NE', 'SW', 'SE'):
                self.matrix.view_scroll(self.dir[state.control])
            elif state.control == 'X':
                self.matrix.view_reset()
            elif state.control == 'Reset':
                self.matrix.reset_matrix()
            elif state.control == 'Clear':
                self.matrix.clear_matrix()
            elif state.control == 'Edit':
                if state.controls[state.control].is_activated():
                    self.matrix.edit_matrix()
                else:
                    self.matrix.activate()
            elif state.control == 'Patterns':
                if self.matrix.patterns:
                    self.ctrl_patterninfo.set_value(self.matrix.patterns[int(state.value.split(' ')[-1])])
                else:
                    self.ctrl_patterns.set_activated(False)
            elif state.control == 'PatternLoad':
                self.load_pattern()
            elif state.control == 'PatternGet':
                self.retrieve_pattern()
            elif state.control == 'Textbox':
                pygame.display.textarea.toggle()
            elif state.control == 'Rate':
                self.matrix.set_tick(int(state.value.split(' ')[-1]))
            elif state.control in ('__Fix', '__TopFix'):
                self.set_moveable('Fixed')
                self.panel_up = not self.panel_up
                if self.panel_up:
                    self.ctrl_topfix.set_enabled(False)
                else:
                    self.ctrl_topfix.set_enabled(True)
            elif state.control == '__Help':
                self.set_tips_display()
            self.matrix.panel_update = True
        if state.panel_update:
            self.matrix.panel_update = True

